### PR TITLE
Cfy 1770 retry mechanism

### PR DESCRIPTION
**Pull request by yaelpeled**

### CODE REVIEW

File: softlayer_plugin/server.py [hunk 1]
```diff
@@ -151,10 +154,39 @@ def _get_vm_details_for_print(details=None):
...
-    active_transaction = virtual_guest_service.getActiveTransaction(
-        id=ctx.instance.runtime_properties[SOFT_LAYER_INSTANCE_ID],
-        mask='transactionGroup')
-    return active_transaction
+    try:
+        return virtual_guest_service.getActiveTransaction(
+            id=ctx.instance.runtime_properties[SOFT_LAYER_INSTANCE_ID],
+            mask='transactionGroup')
+    except SoftLayerAPIError:
+        return ''
```
Comment (yaelpeled) [inline softlayer_plugin/server.py]: check exception

File: softlayer_plugin/server.py [hunk 2]
```diff
@@ -369,36 +367,56 @@ def create(
...
+    ctx.instance.runtime_properties[SOFT_LAYER_INSTANCE_ID] = instance_id
     ctx.instance.runtime_properties[SOFT_LAYER_HOST_NAME] = hostname
 
+
+@operation
+@with_softlayer_client
+def create(
+        softlayer_client,
+        sample_interval_secs,
+        **_):
```
Comment (yaelpeled) [inline softlayer_plugin/server.py]: can be one line

File: softlayer_plugin/server.py [hunk 3]
```diff
@@ -369,36 +367,56 @@ def create(
...
+
+    if ctx.operation.retry_number == 0:
+        _create(softlayer_client)
+
```
Comment (yaelpeled) [inline softlayer_plugin/server.py]: use runtime property instead retry_number=0

File: softlayer_plugin/tests/resources/simple-blueprint.yaml [hunk 4]
```diff
@@ -1,15 +1,8 @@
-# DSL version, should appear in the main blueprint.yaml
-# and may appear in other imports. In such case, the versions must match
 tosca_definitions_version: cloudify_dsl_1_0
 
 imports:
-    # importing cloudify related types, plugins, workflow, etc...
-    # to speed things up, it is possible downloading this file,
-    # including it in the blueprint directory and importing it
-    # instead.
     - http://www.getcloudify.org/spec/cloudify/3.1/types.yaml
```
Comment (yaelpeled) [inline softlayer_plugin/tests/resources/simple-blueprint.yaml]: change to 3.2

File: softlayer_plugin/server.py [hunk 5]
```diff
@@ -369,36 +369,54 @@ def create(
...
+    instance_id = order_result['orderDetails']['virtualGuests'][0]['id']
+    ctx.instance.runtime_properties[SOFT_LAYER_INSTANCE_ID] = instance_id
     ctx.instance.runtime_properties[SOFT_LAYER_HOST_NAME] = hostname
 
+
+@operation
+@with_softlayer_client
+def create(softlayer_client, sample_interval_secs, **_):
+
+    if not ctx.instance.runtime_properties.get(SOFT_LAYER_INSTANCE_ID):
```
Comment (idanmo) [inline softlayer_plugin/server.py]: Please change to: if SOFT_LAYER_INSTANCE_ID in ctx.instance.runtime_properties....

File: softlayer_plugin/server.py [hunk 6]
```diff
@@ -151,10 +154,41 @@ def _get_vm_details_for_print(details=None):
...
-    active_transaction = virtual_guest_service.getActiveTransaction(
-        id=ctx.instance.runtime_properties[SOFT_LAYER_INSTANCE_ID],
-        mask='transactionGroup')
-    return active_transaction
+    try:
+        return virtual_guest_service.getActiveTransaction(
+            id=ctx.instance.runtime_properties[SOFT_LAYER_INSTANCE_ID],
+            mask='transactionGroup')
+    except SoftLayerAPIError as e:
+        if e.faultCode == 'SoftLayer_Exception_ObjectNotFound':
```
Comment (idanmo) [inline softlayer_plugin/server.py]: constant for fault code

**Pull request merged**